src/algebra/algebra/basic.lean

```diff
@@ -1430,18 +1451,27 @@ instance [I : add_comm_group M] : add_comm_group (restrict_scalars R A M) := I
 instance restrict_scalars.module_orig [semiring A] [add_comm_monoid M] [I : module A M] :
   module A (restrict_scalars R A M) := I
 
-variables [comm_semiring R] [semiring A] [algebra R A]
-variables [add_comm_monoid M] [module A M]
+variables [semiring A] [add_comm_monoid M] [module A M]
 
-/--
-When `M` is a module over a ring `A`, and `A` is an algebra over `R`, then `M` inherits a
-module structure over `R`.
-
-The preferred way of setting this up is `[module R M] [module A M] [is_scalar_tower R A M]`.
--/
-instance : module R (restrict_scalars R A M) :=
+/-- When `M` is a module over a semiring `A`, and `A` is an algebra over `R`, then `M` inherits a
+module structure over `R`. -/
+instance [comm_semiring R] [algebra R A] : module R (restrict_scalars R A M) :=
```

Comment:
In the spirit of instances like `submodule.has_scalar`, if the following change is easy to apply, it might mean we only have to apply `restrict_scalars` once.
```suggestion
instance [comm_semiring R] [comm_semiring S] [algebra S R] [algebra R A] [algebra S A] [is_scalar_tower S R A] : 
  module S (restrict_scalars R A M) :=
```

Reply:
Hmm, I think it's not so easy to think of a situation where we'd use this but I guess it can't hurt if it works. I've just added a commit to see :-)

Reply:
This built successfully but raised a linter error. This is easily fixed but made me think about this a bit more and I feel it's just tilting from useful generalisation into non-trivial complication so I've reverted this change, especially since we don't have an application to hand.